src/test/interface/save_metric_json_test.cpp

```diff
@@ -0,0 +1,102 @@
+  std::vector<std::string> output_csv;
+  std::vector<std::string> output_metric;
+  std::vector<std::string> output_metric_4;
+  std::vector<std::string> simplex_model;
+};
+
+TEST_F(CmdStan, save_diag_metric) {
+  std::stringstream ss;
+  ss << convert_model_path(simplex_model) << " random seed=1234"
+     << " method=sample adapt save_metric=1"
+     << " output file=" << convert_model_path(output_csv) << " 2>&1";
+  std::string cmd = ss.str();
+  run_command_output out = run_command(cmd);
+  ASSERT_FALSE(out.hasError);
+  ASSERT_TRUE(file_exists(convert_model_path(output_metric)));
+
+  std::fstream result_stream(convert_model_path(output_metric));
+  std::stringstream result_sstream;
+  result_sstream << result_stream.rdbuf();
+  result_stream.close();
+  std::string metric = result_sstream.str();
+  rapidjson::Document document;
+  ASSERT_FALSE(document.Parse<0>(metric.c_str()).HasParseError());
+  EXPECT_EQ(count_matches("stepsize", metric), 1);
+  EXPECT_EQ(count_matches("inv_metric", metric), 1);
+  EXPECT_EQ(count_matches("[", metric), 1);  // diagonal metric
+}
+
+TEST_F(CmdStan, save_dense_metric) {
+  std::stringstream ss;
+  ss << convert_model_path(simplex_model) << " random seed=1234"
+     << " method=sample adapt save_metric=1 algorithm=hmc metric=dense_e"
+     << " output file=" << convert_model_path(output_csv) << " 2>&1";
+  std::string cmd = ss.str();
+  run_command_output out = run_command(cmd);
+  ASSERT_FALSE(out.hasError);
+  ASSERT_TRUE(file_exists(convert_model_path(output_metric)));
+
+  std::fstream result_stream(convert_model_path(output_metric));
+  std::stringstream result_sstream;
+  result_sstream << result_stream.rdbuf();
+  result_stream.close();
+  std::string metric = result_sstream.str();
+  rapidjson::Document document;
+  ASSERT_FALSE(document.Parse<0>(metric.c_str()).HasParseError());
+  EXPECT_EQ(count_matches("stepsize", metric), 1);
+  EXPECT_EQ(count_matches("inv_metric", metric), 1);
+  EXPECT_EQ(count_matches("[", metric), 7);  // diagonal metric
+}
+
+TEST_F(CmdStan, save_unit_metric) {
+  std::stringstream ss;
+  ss << convert_model_path(simplex_model) << " random seed=1234"
+     << " method=sample adapt save_metric=1 algorithm=hmc metric=unit_e"
+     << " output file=" << convert_model_path(output_csv) << " 2>&1";
+  std::string cmd = ss.str();
+  run_command_output out = run_command(cmd);
+  ASSERT_FALSE(out.hasError);
+  ASSERT_TRUE(file_exists(convert_model_path(output_metric)));
+
+  std::fstream result_stream(convert_model_path(output_metric));
+  std::stringstream result_sstream;
+  result_sstream << result_stream.rdbuf();
+  result_stream.close();
+  std::string metric = result_sstream.str();
+  rapidjson::Document document;
+  ASSERT_FALSE(document.Parse<0>(metric.c_str()).HasParseError());
+  EXPECT_EQ(count_matches("stepsize", metric), 1);
+  EXPECT_EQ(count_matches("inv_metric", metric), 1);
+  EXPECT_EQ(count_matches("[", metric), 1);  // diagonal metric
```

Comment:
```suggestion
  EXPECT_EQ(count_matches("[", metric), 1);  // unit metric is diagonal
```

Reply:
In this case you also know what the metric would be, which you could test if you wanted

Reply:
yes, but string match won't work and parsing out the values from JSON is a PITA.